syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+	s.cancelFunc()
+}
+
+// AwaitSynced returns the channel that will be notified whenever the node enters synced state.
+func (s *Syncer) AwaitSynced() chan struct{} {
+	s.awaitSyncedMu.Lock()
+	defer s.awaitSyncedMu.Unlock()
+	return s.awaitSyncedCh
+}
+
+// ListenToGossip returns true if the node is listening to gossip for blocks/TXs/ATXs data.
+func (s *Syncer) ListenToGossip() bool {
+	return s.conf.AlwaysListen || s.getSyncState() >= gossipSync
+}
+
+// IsSynced returns true if the nodes is fully in sync with the network.
+func (s *Syncer) IsSynced(ctx context.Context) bool {
+	res := s.getSyncState() == synced
+	s.logger.WithContext(ctx).With().Info("node sync state",
+		log.Bool("synced", res),
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+	return res
+}
+
+// Start starts the main sync loop that tries to sync data for every SyncInterval.
+func (s *Syncer) Start(ctx context.Context) {
+	s.syncOnce.Do(func() {
+		if s.ticker.GetCurrentLayer() == 0 {
+			s.setSyncState(ctx, synced)
+		}
+		for {
+			select {
+			case <-s.shutdownCtx.Done():
+				s.logger.WithContext(ctx).Info("stopping sync to shutdown")
+				return
+			case <-s.syncTimer.C:
+				s.synchronize(ctx)
+			}
+		}
+	})
+}
+
+// ForceSync manually start a sync process outside the main sync loop. If the node is already running a sync process,
+// ForceSync will be ignored.
+func (s *Syncer) ForceSync(ctx context.Context) {
+	s.logger.WithContext(ctx).Debug("executing ForceSync")
+	go s.synchronize(ctx)
+}
+
+func (s *Syncer) isClosed() bool {
+	select {
+	case <-s.shutdownCtx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *Syncer) getSyncState() syncState {
+	return (syncState)(atomic.LoadUint32((*uint32)(&s.syncState)))
+}
+
+func (s *Syncer) setSyncState(ctx context.Context, newState syncState) {
+	oldState := syncState(atomic.SwapUint32((*uint32)(&s.syncState), uint32(newState)))
+	if oldState != newState {
+		s.logger.WithContext(ctx).With().Info("sync state change",
+			log.String("from", oldState.String()),
+			log.String("to", newState.String()),
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		events.ReportNodeStatusUpdate()
+		// notify subscribes
+		s.awaitSyncedMu.Lock()
+		defer s.awaitSyncedMu.Unlock()
+		if newState == synced {
+			close(s.awaitSyncedCh)
+		} else if oldState == synced {
+			s.awaitSyncedCh = make(chan struct{})
+		}
+	}
+}
+
+// startNewSyncProcess returns true if syncer is not currently synchronizing.
+func (s *Syncer) startNewSyncProcess() bool {
+	return atomic.CompareAndSwapUint32(&s.isBusy, 0, 1)
+}
+
+func (s *Syncer) endSyncProcess() {
+	atomic.StoreUint32(&s.isBusy, 0)
+}
+
+func (s *Syncer) setTargetSyncedLayer(ctx context.Context, layerID types.LayerID) {
```

Comment:
add a comment here as well to what a "target sync layer" is

Reply:
done